guide: reorder how-tos and copy edits; expand on adding existing dependencies to a stage (#1914)
iterative/dvc.org: content/docs/user-guide/how-to/add-deps-or-outs-to-a-stage.md

# Add Dependencies or Outputs to a Stage

There are situations where we have executed a stage, but later notice that some
of the files/directories used by the stage as dependencies or created as outputs
are missing from `dvc.yaml`.

To add existing files/directories as <abbr>dependencies</abbr> or

> <abbr>outputs</abbr> to a stage, either edit the `dvc.yaml` file or use

Author: to a stage "without executing it again" ?

@imhardikj

`dvc run` with the `--no-exec` option. Then use `dvc commit` to save the

> output(s) to the <abbr>cache</abbr> (and update `dvc.lock`).

Author: WDYT @shcheklein ? It's still specific to existing dep/out files or dirs. (which was our original intention). Do you think we need to generalize it even more and leave the `commit` part as an optional step?

Either way I say this is ready to be merged for now, but since I opened the PR, I can't approve it. If you agree, please do.

Reviewer: > It's still specific to existing dep/out files or dirs (which was our original intention)

then name should be different?

the point is- name is general, content is not. it makes a bit misleading. Like people might think that this the only way to add, or that it's the most general way when add is needed, etc, etc

Author: OK so @imhardikj, to generalize this we need to:
- Invert the order of the paragraphs above (don't talk about existing files in the first one)
- The `dvc commit` part should be in the 2nd one
- Edit text accordingly so it makes sense after the change ^
- The example will also need to emphasize that the last part is needed only for existing deps/outs files or dirs ("Finally, we need...")

Reviewer 2: @jorgeorpinel updated with changes as per your suggestion.

Author: Hmmm yeah it didn't make much sense. Bullets 3 and 4 weren't really addressed. I'm going to take over this part... Thanks, resolving.

# Add Dependencies or Outputs to a Stage

> There are situations where we have executed a stage (either by writing

Reviewer: In this case title sounds too general for the `dvc commit` scenario. It sounds like all we need to describe in this section is `open dvc.yaml - edit` or execute `dvc run` again. If it happened that files are already in place use --no-exec + `dvc commit`

Reviewer 2: In both cases (either editing `dvc.yaml` or using `--no-exec` with `run`), `dvc commit` is required. So that's why it's a separate scenario in `dvc commit`.

Author: > too general for the dvc commit scenario... all we need to describe in this section is...

@shcheklein do you mean in https://github.com/iterative/dvc.org/pull/1914/files#diff-b8f371cb91c101e733f37e95de023afe203b77b8a0443b3d453a76739b675852R43-R48 above?

The scenario in `commit` is "we have previously executed a stage and later notice that some files/directories are missing from `deps` or `outs` in `dvc.yaml`" which seems to me exactly like what this doc addresses.

> Cc @imhardikj that would be the whole text in it's most compressed form BTW.

Or are you saying that the long text and example in this how-to is unnecessary and could be compressed into a single paragraph (thus maybe not even needing a whole how-to page)?

Thanks

Reviewer: My point is that when I read `Add Dependencies or Outputs to a Stage` I would expect to a much simpler scenario first - `Open your file, add it manually, save file. Run repro or commit in some cases.` - that kind of a thing. Not one of the very specific sub-scenarios from the `dvc commit`.

Author: Gotcha Ivan. You're right. We should simplify the intro to this doc and separate that from the example below @imhardikj. I'll resolve this conversation and reopen in the exact places...

`dvc.yaml` manually and using `dvc repro`, or with `dvc run`), but later notice
that some of its data requirements are missing from `dvc.yaml`. Namely, one of:

- Files or directories in the <abbr>workspace</abbr> that are dependencies of
  the stage are missing from `deps`.
- Output files or directories that the stage creates, which are already in the
  workspace are missing from `outs`.

Follow the steps below to add existing files/directories as
<abbr>dependencies</abbr> or <abbr>outputs</abbr> to a stage without executing

> it again, which can be expensive/time-consuming, and is unnecessary.

Author: Ok so this can be simplified quite a bit probably, as mentioned in https://github.com/iterative/dvc.org/pull/1914#pullrequestreview-534022299. E.g.:

> all we need to describe in this section is open dvc.yaml - edit or execute dvc run again. If it happened that files are already in place use --no-exec + dvc commit

> much simpler scenario and recommendation first - Open your file, add it manually, save file. Run repro or commit in some cases.

Reviewer 2: Simplified this into two small paragraphs.

Author: Very good, thanks.

We start with an example `prepare` stage, which has a single dependency and
output. To add a missing dependency (`data/raw.csv`) as well as a missing output

> (`data/validate`) to this stage, we can edit `dvc.yaml` like this:

Author: And starting somewhere around here it can all be in a separate `## Example` section.

Reviewer 2: Added `## Example` section.

> # How to Add Dependencies or Outputs

Reviewer: add to something?

Author: Unfortunately the title is too long so I felt like we needed to compromise. See nav vs. title here:

![image](https://user-images.githubusercontent.com/1477535/100289387-5bd38a00-2f3e-11eb-96a6-419b5df9dbb8.png)

I've added a SEO `title` now that is more correct though. Should help

Author: p.s. on this, I think it's OK since "Dependencies or Outputs" can only apply to stages anyway.

> We have executed a stage, but later notice that some of the files/directories it

Reviewer: Is it needed? the whole paragraph?

Author: It's a single sentence on motivation.

Like I mentioned, I realized that this how-to is really only useful for people who have already reproed their dvc.yaml or used run from the beginning. For people that write dvc.yaml manually and haven't reproed it, it's obvious how to add more deps/outs.

Author: > this how-to is really only useful for people who have already reproed their dvc.yaml or used run from the beginning

Let me think this through thouogh... ⌛

Author: OK I went your way on this. PTAL

uses as dependencies or creates as outputs are missing from `dvc.yaml`...

To add files or directories as <abbr>dependencies</abbr> or <abbr>outputs</abbr>

> to an existing stage without re-executing it (which can be expensive and is

Reviewer: I would still simplify it:

To add files or directories ... edit the `dvc.yaml` (alternatively do `dvc run --no-exec` if you still have this command around).

Now, to update the `dvc.lock` either run `dvc repro` or `dvc commit` to "force" it in case outputs already exist and you don't want to run anything again.

-----

It's better to push people editing file manually. `dvc run` is complicated, people don't have it around sometimes, etc.

Author: > to update the dvc.lock either run dvc repro

But the scenario is "add deps/outs to an existing stage **without re-executing it**". I simplified it along the lines of your suggestion for now, but skipped the parts that contradict that.

  the stage are missing from `deps`.

> 

Author: I thought I removed this line before. Is the formatting inserting it, or is that you @imhardikj? (Please see https://dvc.org/doc/user-guide/contributing/docs#doc-style-guidelines-javascript-and-markdown about bullet lists.) 
```suggestion
```

Reviewer 2: Actually it's my mistake. I but didn't notice that you removed the empty line in f578ddd06f32b6b660a5ff5917658d771ca76474. Updated now.

Author: No problem. Thanks

> with the `-f` and `--no-exec` options to the same end.

Author: I simplified and generalized this new one as we agreed recently.

>   workspace, are missing from `outs` field.

Author: "from deps" or "from THE deps field"
Same for "outs"

Author: Let's just just say `deps`/`outs` (no "field").

> After the update, use `dvc commit` to save any output files that already exist

Reviewer: This should go as a very specific note if it's a general case. In general case you edit, and run `dvc repro`. `dvc commit` should be considered as a hack in most cases. Users should understand what are they doing when they use it (potentially break reproducibility).

Author: OK, I started with a conditional now, but I don't imagine this is such a rare edge case. May even be the norm that you realize the stage needs an update only after haven forgotten to define all the outs and executing it that way.

I don't know if the generic case of just continuing to edit dvc.yaml before having reproed the pipeline even needs a doc. It will be pretty obvious for those who create stages by hand by editing dvc.yaml, no? People who may find this useful will have already done a repro or used run from the beginning, I assume.

Follow the steps below to add existing files/directories as
<abbr>dependencies</abbr> or <abbr>outputs</abbr> to a stage without

> re-executing it again, which can be expensive/time-consuming, and is

Author: "re" + "again" is redundant:
```suggestion
executing it again, which can be expensive/time-consuming, and is
```

it again, which can be expensive/time-consuming, and is unnecessary.

> We start with an example stage `prepare`, which has a single dependency and

Author: "exapmle stage prepare"? Sounds strange.
```suggestion
We start with an example `prepare` stage , which has a single dependency and
```

Reviewer 2: Updated

> We start with an example `prepare`, which has a single dependency and output. To

Author: Example `prepare` what?

> output. To add a missing dependency `data/raw.csv` and output `data/validate` to

Author: Let's be clearer, since this part got complicated:
```suggestion
output. To add a missing dependency (`data/raw.csv`) as well as a missing output (`data/validate`) to
```

Reviewer 2: Done

> add a missing dependency `data.csv`, and output `data/validate` to this stage,

Author: I'm starting to notice your use of commas is sometimes unnecessary. Not a big deal and sometimes stylistic, but please keep an eye on that @imhardikj. Thanks:
```suggestion
add a missing dependency `data.csv` and output `data/validate` to this stage,
```

we can edit `dvc.yaml` like this:

```git
 stages:
   prepare:
     cmd: python src/prepare.py
     deps:

> +    - data.csv

Author: Should probably be a path inside `data/`?
```suggestion
+    - data/raw.csv
```

Author: Applies in a few other places.

# Add Dependencies or Outputs to a Stage

There are situations where we have executed a stage (either by writing
`dvc.yaml` manually and using `dvc repro`, or with `dvc run`), but later notice

> that some of the build requirements are missing from `dvc.yaml`:

Author: Good idea. Maybe something like is better though:
```suggestion
that some of its data requirements are missing from `dvc.yaml`. Namely, one of:
```
May need formatting

- Files or directories in the <abbr>workspace</abbr> that are dependencies of

>   the stage, are missing from `deps` field.

Reviewer 2: @jorgeorpinel I have made some changes to add dependencies part. Also updated to run/commit.

Author: Sure, reviewing https://github.com/iterative/dvc.org/pull/1913#pullrequestreview-525917714 now. Resolving this. Thanks

>   the stage, are missing from `deps` field.
> 

Author: ```suggestion
  the stage are missing from `deps` field.
```
